From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Juha Kallioinen <juha.kallioinen@jolla.com>
Date: Wed, 6 Nov 2013 14:05:27 +0000
Subject: [PATCH] Enable netrc-optional on libcurl to allow for easier image
 building in SDK

Original patch from: Islam Amer <islam.amer@jollamobile.com>

Signed-off-by: Juha Kallioinen <juha.kallioinen@jolla.com>
---
 zypp/media/MediaCurl.cc | 58 ++++++++++++++++++++++++++++++++---------
 zypp/media/MediaCurl.h  |  2 +-
 2 files changed, 47 insertions(+), 13 deletions(-)

diff --git a/zypp/media/MediaCurl.cc b/zypp/media/MediaCurl.cc
index 3f310688c9da252d7b1a32f2c1459805d3db135f..921b9f09275e0e29620d1b266d00e23bc0358792 100644
--- a/zypp/media/MediaCurl.cc
+++ b/zypp/media/MediaCurl.cc
@@ -749,9 +749,10 @@ void MediaCurl::getFileCopy( const OnMediaLocation & srcFile , const Pathname &
   Url fileurl(getFileUrl(filename));
 
   bool firstAuth = true;  // bsc#1210870: authenticate must not return stored credentials more than once.
+  bool netrc = false;
+  int numTry = 0;
   unsigned internalTry = 0;
   static constexpr unsigned maxInternalTry = 3;
-
   do
   {
     try
@@ -762,7 +763,7 @@ void MediaCurl::getFileCopy( const OnMediaLocation & srcFile , const Pathname &
     // retry with proper authentication data
     catch (MediaUnauthorizedException & ex_r)
     {
-      if ( authenticate(ex_r.hint(), firstAuth) ) {
+      if ( authenticate(ex_r.hint(), numTry++, netrc) ) {
         firstAuth = false;  // must not return stored credentials again
         continue; // retry
       }
@@ -801,8 +802,8 @@ void MediaCurl::getFileCopy( const OnMediaLocation & srcFile , const Pathname &
 
 bool MediaCurl::getDoesFileExist( const Pathname & filename ) const
 {
-  bool retry = false;
-
+  bool retry = false, netrc = false;
+  int numTry = 0;
   do
   {
     try
@@ -812,7 +813,7 @@ bool MediaCurl::getDoesFileExist( const Pathname & filename ) const
     // authentication problem, retry with proper authentication data
     catch (MediaUnauthorizedException & ex_r)
     {
-      if(authenticate(ex_r.hint(), !retry))
+      if (authenticate(ex_r.hint(), numTry++, netrc))
         retry = true;
       else
         ZYPP_RETHROW(ex_r);
@@ -1410,9 +1411,32 @@ void MediaCurl::resetExpectedFileSize(void *clientp, const ByteCount &expectedFi
 }
 
 ///////////////////////////////////////////////////////////////////
-
-bool MediaCurl::authenticate(const std::string & availAuthTypes, bool firstTry) const
+/*
+ * The authentication is a challenge-response type transaction. We
+ * come here after the challenge has been received and need to send a
+ * response. There are plenty of ways to send the right and the wrong
+ * response. All of these preconditions need to be considered:
+ *
+ * 1) there are no existing credentials
+ * 2) credential manager has right/wrong credentials
+ * 3) user enters right/wrong credentials interactively
+ * 4) .netrc contains right/wrong credentials
+ * 5) client (e.g. zypper) can be in interactive or non-interactive mode
+ *
+ * First we always want to try to send a response with any stored
+ * credentials. If there are none, then we'll try using a .netrc. Only
+ * after these methods have failed to authenticate the user, we'll
+ * prompt the user for the credentials or give up if in
+ * non-interactive mode.
+ *
+ * The challenge-response loop needs to be able to end in the
+ * non-interactive mode in case none of the available methods provide
+ * the correct response.
+ *
+ */
+bool MediaCurl::authenticate(const std::string & availAuthTypes, int numTry, bool &netrcUsed) const
 {
+  DBG << "numtry: " << numTry << endl;
   //! \todo need a way to pass different CredManagerOptions here
   CredentialManager cm(CredManagerOptions(ZConfig::instance().repoManagerRoot()));
   CurlAuthData_Ptr credentials;
@@ -1420,21 +1444,29 @@ bool MediaCurl::authenticate(const std::string & availAuthTypes, bool firstTry)
   // get stored credentials
   AuthData_Ptr cmcred = cm.getCred(_url);
 
-  if (cmcred && firstTry)
+  // first try with any stored credentials
+  if (cmcred && (numTry == 0))
   {
     credentials.reset(new CurlAuthData(*cmcred));
     DBG << "got stored credentials:" << endl << *credentials << endl;
   }
-  // if not found, ask user
-  else
-  {
+  // no stored creds or they failed, try .netrc instead if not already tried
+  else if ((numTry == 0 || numTry == 1) && (!netrcUsed)) {
+    DBG << "try with .netrc" << endl;
+    CURLcode ret = curl_easy_setopt(_curl, CURLOPT_NETRC, CURL_NETRC_OPTIONAL);
+    if ( ret != 0 ) ZYPP_THROW(MediaCurlSetOptException(_url, _curlError));
+    netrcUsed = true;
+    return true;
+  }
+  // stored creds and .netrc failed, ask user
+  else {
 
     CurlAuthData_Ptr curlcred;
     curlcred.reset(new CurlAuthData());
     callback::SendReport<AuthenticationReport> auth_report;
 
     // preset the username if present in current url
-    if (!_url.getUsername().empty() && firstTry)
+    if (!_url.getUsername().empty() && (numTry == 0))
       curlcred->setUsername(_url.getUsername());
     // if CM has found some credentials, preset the username from there
     else if (cmcred)
@@ -1471,6 +1503,7 @@ bool MediaCurl::authenticate(const std::string & availAuthTypes, bool firstTry)
     }
     else
     {
+      // can be the result of the non-interactive client mode
       DBG << "callback answer: cancel" << endl;
     }
   }
@@ -1509,6 +1542,7 @@ bool MediaCurl::authenticate(const std::string & availAuthTypes, bool firstTry)
     return true;
   }
 
+  // ends the authentication challenge-response loop
   return false;
 }
 
diff --git a/zypp/media/MediaCurl.h b/zypp/media/MediaCurl.h
index 14b56f605bbbf384c1208c2322e88bfeb4eaa870..cb399f5ece0687d4e7ea2b32c77496c29e7bae5e 100644
--- a/zypp/media/MediaCurl.h
+++ b/zypp/media/MediaCurl.h
@@ -152,7 +152,7 @@ class MediaCurl : public MediaNetworkCommonHandler
      */
     std::string getAuthHint() const;
 
-    bool authenticate(const std::string & availAuthTypes, bool firstTry) const;
+    bool authenticate(const std::string & availAuthTypes, int numTry, bool& netrcUsed) const;
 
     bool detectDirIndex() const;
 
